src/pauli_exp_val.rs

```diff
@@ -0,0 +1,193 @@
+        let mut val: f64 = data_arr[i].re * data_arr[i].re + data_arr[i].im * data_arr[i].im;
+        if (i & z_mask).count_ones() & 1 != 0 {
+            val *= -1.;
+        }
+        val
+    };
+
+    if num_qubits < 19 || !run_in_parallel {
+        Ok(fast_sum(&(0..size).map(map_fn).collect::<Vec<f64>>()))
+    } else {
+        Ok((0..size).into_par_iter().map(map_fn).sum())
+    }
+}
+
+/// Compute the pauli expectatation value of a statevector with x
+#[pyfunction]
+#[pyo3(text_signature = "(data, num_qubits, z_mask, x_mask, phase, x_max, /)")]
+pub fn expval_pauli_with_x(
+    data: PyReadonlyArray1<Complex64>,
+    num_qubits: usize,
+    z_mask: usize,
+    x_mask: usize,
+    phase: Complex64,
+    x_max: u32,
+) -> PyResult<f64> {
+    let data_arr = data.as_slice()?;
+    let mask_u = !(2_usize.pow(x_max + 1) - 1);
+    let mask_l = 2_usize.pow(x_max) - 1;
+    let size = 1 << (num_qubits - 1);
+    let run_in_parallel = eval_parallel_env();
+    let map_fn = |i: usize| -> f64 {
+        let index_0 = ((i << 1) & mask_u) | (i & mask_l);
+        let index_1 = index_0 ^ x_mask;
+        let val_0 = (phase
+            * Complex64::new(
+                data_arr[index_1].re * data_arr[index_0].re
+                    + data_arr[index_1].im * data_arr[index_0].im,
+                data_arr[index_1].im * data_arr[index_0].re
+                    - data_arr[index_1].re * data_arr[index_0].im,
+            ))
+        .re;
+        let val_1 = (phase
+            * Complex64::new(
+                data_arr[index_0].re * data_arr[index_1].re
+                    + data_arr[index_0].im * data_arr[index_1].im,
+                data_arr[index_0].im * data_arr[index_1].re
+                    - data_arr[index_0].re * data_arr[index_1].im,
+            ))
+        .re;
+        let mut val = val_0;
+        if (index_0 & z_mask).count_ones() & 1 != 0 {
+            val *= -1.
+        }
+        if (index_1 & z_mask).count_ones() & 1 != 0 {
+            val -= val_1;
+        } else {
+            val += val_1;
+        }
+        val
+    };
+    if num_qubits < 19 || !run_in_parallel {
+        Ok(fast_sum(&(0..size).map(map_fn).collect::<Vec<f64>>()))
+    } else {
+        Ok((0..size).into_par_iter().map(map_fn).sum())
+    }
+}
+
+/// Compute the pauli expectatation value of a density matrix without x
+#[pyfunction]
+#[pyo3(text_signature = "(data, num_qubits, z_mask, /)")]
+pub fn density_expval_pauli_no_x(
+    data: PyReadonlyArray1<Complex64>,
+    num_qubits: usize,
+    z_mask: usize,
+) -> PyResult<f64> {
+    let data_arr = data.as_slice()?;
+    let num_rows = 1 << num_qubits;
+    let stride = 1 + num_rows;
+    let run_in_parallel = eval_parallel_env();
+    let map_fn = |i: usize| -> f64 {
+        let index = i * stride;
+        let mut val = data_arr[index].re;
+        if (i & z_mask).count_ones() & 1 != 0 {
+            val *= -1.;
+        }
+        val
+    };
+    if num_qubits < 19 || !run_in_parallel {
```

Comment:
Reminder for pulling `19` out into a constant, should you decide to do it!

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7702/commits/44efb30f3af504677d0d45dc2946e934d980af80